### PR TITLE
-n option and fix for .only on dup test names

### PR DESCRIPTION
**Pull request by jtlapp**
This PR provides the `-n` option described in #301 and the fix for #299 on which it depends.

### CODE REVIEW

File: bin/tape [hunk 1]
```diff
@@ -25,6 +26,17 @@ opts.require.forEach(function(module) {
     }
 });
 
+if (opts.number !== false) {
+    var tape;
+    try {
+        tape = require(resolveModule("tape", { basedir: cwd }));
+    }
+    catch(err) {
+        tape = require("../"); // assume testing tape module itself
```
Comment (ljharb) [inline bin/tape]: can you elaborate on this code block? adding production code just for tests seems very subpar.

File: index.js [hunk 2]
```diff
@@ -119,8 +120,11 @@ function createHarness (conf_) {
                 if (!r.ok && typeof r !== 'string') test._exitCode = 1
             });
         })(t);
-        
+    
         results.push(t);
+        if (t.number === onlyTestNumber)
+            results.only(onlyTestNumber);
```
Comment (ljharb) [inline index.js]: please add curly braces around all if blocks, in this case for consistency with lines 107-109.

File: index.js [hunk 3]
```diff
@@ -138,13 +142,23 @@ function createHarness (conf_) {
     var only = false;
     test.only = function (name) {
         if (only) throw new Error('there can only be one only test');
-        results.only(name);
         only = true;
-        return test.apply(null, arguments);
+        var t = test.apply(null, arguments);
+        if (!onlyTestNumber)
+            results.only(t.number);
```
Comment (ljharb) [inline index.js]: please add curly braces around all if blocks, in this case for consistency with lines 107-109.

File: index.js [hunk 4]
```diff
@@ -138,13 +142,23 @@ function createHarness (conf_) {
...
     return test;
 }
+
+var onlyTestNumber = false;
+
+function setNumbering (number) {
+    Test.showTestNumbers();
+    if (typeof number === "number")
+        onlyTestNumber = number;
+}
```
Comment (ljharb) [inline index.js]: please ensure all files have a trailing newline

File: lib/test.js [hunk 5]
```diff
@@ -44,8 +46,11 @@ function Test (name_, opts_, cb_) {
 
     var args = getTestArgs(name_, opts_, cb_);
 
+    this.number = nextTestNumber++;
     this.readable = true;
     this.name = args.name || '(anonymous)';
+    if (numbered)
+        this.name = '['+ this.number +'] '+ this.name;
```
Comment (ljharb) [inline lib/test.js]: please add curly braces around all if blocks

File: lib/test.js [hunk 6]
```diff
@@ -499,5 +504,9 @@ Test.skip = function (name_, _opts, _cb) {
     return Test(args.name, args.opts, args.cb);
 };
 
+Test.showTestNumbers = function() {
```
Comment (ljharb) [inline lib/test.js]: please add a space after "function"

**Comment by jtlapp** [comment]
Oops, forgot to check for consistency with existing style.

Also, my colleague found a bug whereby indicating a test number greater than the number of tests results in all tests running. I'm thinking that I should either run no tests in this case and do one of the following:

- output a message to the console (stdout, I presume)
- output a message to stderr
- output nothing

**Comment by ljharb** [comment]
I think that should error out without running any tests, yes.

What if I run a number that's explicitly skipped? I assume it would still be skipped.

**Comment by jtlapp** [comment]
I agree, it should still be skipped. I'll make some more unit tests. Thank you!

**Comment by jtlapp** [comment]
Another to-do item for me: Make sure the bracketed numbers are compatible with the TAP standard.

**Comment by jtlapp** [comment]
When I've got this done, should I close this PR and make a new one? (I'm new to git/github.)

**Comment by ljharb** [comment]
@jtlapp no, please do not. Rebase and force push and the PR will be updated.

**Comment by jtlapp** [comment]
Okay, I'll try to make sense of that. Out of curiosity, why a trailing newline?

**Comment by ljharb** [comment]
http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

**Comment by jtlapp** [comment]
Thanks. That explains why I have a habit of ending all my text/data files in a newline, but I guess it's been a while since I've catted or munged code via command line tools.

File: bin/tape (hunk 1 above)
Comment (jtlapp) [inline bin/tape]: By way of explanation, the `-n` option must inform the tape module of the number of the test to run. The above `resolveModule()` call requires that a module named `tape` have been installed. No module named `tape` has been installed when we run the test suite for `tape` itself. In order to be able to test the `-n` option within the `tape` module itself, it's necessary to load the module in the way that is required by the test suite. You'll find all the tests loading the `tape` module this way as well (not just mine). Without this line, the `tape` module would not be able to include tests of the `-n` option.

Comment (ljharb) [inline bin/tape]: It seems like we could resolve that by adding a devDep of `"tape": "./"`?

Comment (jtlapp) [inline bin/tape]: I don't know. That's trick NPM stuff. That might allow you to replace the mysterious `var tape = require('../');` line found in each test file.

**Comment by jtlapp** [comment]
Presently, if you use `-n` for a test that is being skipped, you get the same result as doing this using your current code base:

`test.only("skip or only?", {skip: true}, (t) => { ... }`

Which is this:

    # SKIP lexically invalid domain name
    
    1..0
    # tests 0
    # pass  0
    
    # ok

The numbering shows that a test is being skipped -- numbers are skipped. I like that. It's informative.

I'm getting my `-n` implementation right before looking for how to hook it in.

File: bin/tape (hunk 1 above)
Comment (ljharb) [inline bin/tape]: That line isn't mysterious - that's how every npm package tests itself.

**Comment by jtlapp** [comment]
I'm closing this. I found an easy way to extend the `tap` module to do the job. Thanks for your help!

**Pull request closed without merging**